lib/rubocop/cop/rails/i18n_locale_texts.rb

```diff
@@ -0,0 +1,109 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Rails
+      # Enforces use of I18n and locale files instead of locale specific strings.
+      #
+      # @example
+      #   # bad
+      #   class Product < ApplicationRecord
+      #     validates :code, format: { with: /\A[a-zA-Z]+\z/, message: "only letters" }
+      #   end
+      #
+      #   # good
+      #   # config/locales/en.yml
+      #   # en:
+      #   #   activerecord:
+      #   #     errors:
+      #   #       models:
+      #   #         product:
+      #   #           code_format: "only letters"
+      #
+      #   class Product < ApplicationRecord
+      #     validates :code, format: { with: /\A[a-zA-Z]+\z/, message: :code_format }
+      #   end
+      #
+      #   # bad
+      #   class PostsController < ApplicationController
+      #     def create
+      #       # ...
+      #       redirect_to root_path, notice: "Post created!"
+      #     end
+      #   end
+      #
+      #   # good
+      #   # config/locales/en.yml
+      #   # en:
+      #   #   posts:
+      #   #     create:
+      #   #       success: "Post created!"
+      #
+      #   class PostsController < ApplicationController
+      #     def create
+      #       # ...
+      #       redirect_to root_path, notice: t(".success")
+      #     end
+      #   end
+      #
+      #   class UserMailer < ApplicationMailer
+      #     def welcome(user)
+      #       mail(to: user.email, subject: "Welcome to My Awesome Site")
+      #     end
+      #   end
+      #
+      #   # good
+      #   # config/locales/en.yml
+      #   # en:
+      #   #   user_mailer:
+      #   #     welcome:
+      #   #       subject: "Welcome to My Awesome Site"
+      #
+      #   class UserMailer < ApplicationMailer
+      #     def welcome(user)
+      #       mail(to: user.email)
+      #     end
+      #   end
+      #
+      class I18nLocaleTexts < Base
+        MSG = 'Move locale texts to the locale files in the `config/locales` directory.'
+
+        RESTRICT_ON_SEND = %i[validates redirect_to []= mail].freeze
+
+        def_node_search :validation_message, <<~PATTERN
+          (pair (sym :message) $str_type?)
+        PATTERN
+
+        def_node_search :redirect_to_flash, <<~PATTERN
+          (pair (sym {:notice :alert}) $str_type?)
+        PATTERN
+
+        def_node_matcher :flash_assignment?, <<~PATTERN
+          (send (send nil? :flash) :[]= _ $str_type?)
+        PATTERN
+
+        def_node_search :mail_subject, <<~PATTERN
+          (pair (sym :subject) $str_type?)
+        PATTERN
+
+        def on_send(node)
+          case node.method_name
+          when :validates
+            validation_message(node).each do |text_node|
```

Comment:
Cosmetic: is `.each` necessary? Would just `validation_message(node) do |text_node|` work?
I understand that `def_node_search` needs to be replaced with `def_node_matcher`, and the node pattern gains more weight.

I'm thinking of something like (please accept my apologies for a contrived example):
```ruby
  validates :username,
    uniqueness: {
      message: ->(object, data) do
        {message: 'ough oh', value: data.value}.to_s
      end
    }
```

Reply:
Writing this as:
```ruby
def on_send(node)
  validation_message?(node, &method(:add_offence))
  redirect_to_flash?(node, &method(:add_offence))
  flash_assignment?(node, &method(:add_offence))
  mail_subject?(node, &method(:add_offence))
end
```
would also allow getting rid of an early return and `to_a.last`.

Subjectively, cleaner. But again, pretty cosmetic.

Reply:
> Cosmetic: is .each necessary?

Checked, indeed works without it.

Agreed, cleaner, despite I personally not a fan of `&method` construction.
But the possible problem with this is that all the lines are executed, even if we have validation and it should skip the rest of the checks.

So, should be done as `return if something?(...)` x 4 ?